src/Controls/src/SourceGen/CodeBehindGenerator.cs

```diff
@@ -364,24 +388,49 @@ static string GetTypeName(XmlType xmlType, Compilation compilation, IList<XmlnsD
 			return XmlnsHelper.ParseNamespaceFromXmlns(namespaceuri);
 		}
 
-		static string GetTypeNameFromCustomNamespace(XmlType xmlType, Compilation compilation, IList<XmlnsDefinitionAttribute> xmlnsDefinitionCache)
+		static string GetTypeNameFromCustomNamespace(XmlType xmlType, Compilation compilation, AssemblyCaches caches)
 		{
 #nullable disable
-			string typeName = xmlType.GetTypeReference<string>(xmlnsDefinitionCache, null,
+			string typeName = xmlType.GetTypeReference<string>(caches.XmlnsDefinitions, null,
 				(typeInfo) =>
 				{
 					string typeName = typeInfo.typeName.Replace('+', '/'); //Nested types
 					string fullName = $"{typeInfo.clrNamespace}.{typeInfo.typeName}";
+					IList<INamedTypeSymbol> types = compilation.GetTypesByMetadataName(fullName);
+
+					if (types.Count == 0)
+						return null;
+
+					foreach (INamedTypeSymbol type in types)
+					{
+						// skip over types that are not in the correct assemblies
+						if (type.ContainingAssembly.Identity.Name != typeInfo.assemblyName)
+							continue;
+
+						if (IsPublicOrVisibleInternal(type, caches.InternalsVisible))
+							return fullName;
+					}
```

Comment:
This is the real fix for the code generator. Instead of failing with duplicate types, we iterate over them and see which one is the visible one.